docs/development/design/github-application.rst

```diff
@@ -0,0 +1,306 @@
+
+
+GitHub Application requirements
+-------------------------------
+
+:Repository permissions:
+   - Metadata: Read-only (mandatory)
+   - Contents: Read-only
+   - Commit statuses: Read & Write
+   - Pull requests: Read-only
+
+:Organization permissions:
+   - Members: Read-only
+
+:User permissions:
+   - Email addresses: Read-only
+
+:Events subscribed:
+   - Meta (When this App is deleted and the associated hook is removed)
+   - Create (Branch or tag created)
+   - Delete (Branch or tag deleted)
+   - Push (Git push to a repository)
+   - Pull request (Pull request opened, closed, reopened, ...)
+
+
+Authentication as GitHub Application
+------------------------------------
+
+https://docs.github.com/en/developers/apps/building-github-apps/authenticating-with-github-apps
+
+Generate the JWT on Python:
+
+.. code:: python
+
+   import datetime
+   import jwt
+   GITHUB_APP_ID = 134302
+
+   # content of PEM file downloaded from GH Application settings
+   private_key = b'''-----BEGIN RSA PRIVATE KEY-----'''
+
+   jwt.encode({"iat": datetime.datetime.utcnow() - datetime.timedelta(seconds=60), "exp": datetime.datetime.utcnow() + datetime.timedelta(minutes=5), "iss": GITHUB_APP_ID}, private_key, algorithm="RS256")
+
+Use the token generated in the previous step as authorization in the cURL command:
+
+.. code:: bash
+
+   $ curl -H "Authorization: Bearer <JWT>" -H "Accept: application/vnd.github.v3+json" https://api.github.com/app
+
+
+.. note::
+
+   For some reason this is not working and I'm getting:
+
+   .. code:: text
+
+     'Issued at' claim ('iat') must be an Integer representing the time that the assertion was issued
+
+
+SSH keys are not required to clone private repositories
+-------------------------------------------------------
+
+When asking for ``contents`` (repository permission) we are able to clone
+private repositories by using GitHub Application installation access tokens:
+
+.. code:: bash
+
+   git clone https://x-access-token:<token>@github.com/owner/repo.git
+
+(from
+https://docs.github.com/en/developers/apps/building-github-apps/authenticating-with-github-apps#http-based-git-access-by-an-installation)
+
+
+Handling webhooks
+-----------------
+
+GitHub Application has *only one webhook* where it will receive all the events
+for all the installations. The body contains all the information about the
+particular installation that triggered the event. With this data, we will create
+an access token and perform the query/actions we need.
+
+There are some events that need to map to particular ``Project`` in our
+database. For example, "a push to ``main`` branch in repository
+``readthedocs/blog``" should "trigger a build for ``latest`` version on
+``read-the-docs-blog`` project". For these cases we can use use
+``repository.id`` field from the body to get the ``RemoteRepository.remote_id``
+and from there get the ``Project``.
```

Comment:
This is just a Project mapping though, and means that if `readthedocs/template` gets a commit, then *every* project on RTD with that `repo_url` will be built? That takes us back to our original webhooks implementation, which is not great :/

Reply:
> This is just a Project mapping though, and means that if `readthedocs/template` gets a commit, then _every_ project on RTD with that `repo_url` will be built?

No.

This will trigger a build _only_ on the projects that were imported by users with permissions on `readthedocs/template` repository. That is, it appeared on the "Import Project" page and they clicked on "+" to import it. This is where the relation between `RemoteRepository -> Project` is created.

All the `readthedocs/template` forks and, all official `readthedocs/template` that were imported manually _won't be associated to any `RemoteRepository`_, so it won't trigger a build for those on commits.